File: src/Core.TypeScript/quantum/quantum-arith.ts
/**
 * quantum-arith.ts — Canonical quantum arithmetic library (TypeScript port).
 *
 * This is the byte-lock reference for the TypeScript substrate.
 * All 9 golden vectors must match the F# canonical implementation exactly.
 *
 * ## Hexagonal port pattern
 *
 * The `IQuantumArith` interface is the port. `canonicalArith` is our
 * implementation. `qrispAdapter` and `openQASMAdapter` are stubs for
 * external libraries — they delegate to canonical until the external
 * library is integrated.
 *
 * ## Byte-lock discipline
 *
 * All operations use IEEE 754 double precision (64-bit) with
 * round-to-nearest-even. No FMA, no fast-math, no platform-specific
 * intrinsics. The golden vectors are in `testdata/quantum-arith-golden.json`.
 *
 * ## Golden vectors (QA-1..QA-9)
 *
 * QA-1: complex_add(1+2i, 3+4i) = 4+6i
 *   hex: 4010000000000000 4018000000000000
 * QA-2: complex_mul(1+2i, 3+4i) = -5+10i
 *   hex: c014000000000000 4024000000000000
 * QA-3: complex_mag_sq(3+4i) = 25.0
 *   hex: 4039000000000000
 * QA-4: blaschke(0.5+0.3i, sqrt(λ₀)·e^{iπ/4}) = 0.474586592674752+0.260348313343704i
 *   hex: 3fde5fa071aa1ed9 3fd0a98bf8d8516e
 * QA-5: blaschkeDerivMagSq(0.5+0.3i, sqrt(λ₀)·e^{iπ/4}) = 1.147451008268201
 *   hex: 3ff25bf596a462ea
 * QA-6: invSqrt2 = 0.707106781186547
 *   hex: 3fe6a09e667f3bcc
 * QA-7: bornProb(|Φ⁺⟩, 00) = 0.5
 *   hex: 3fdffffffffffffe
 * QA-8: tsirelsonS = 2.828427124746190
 *   hex: 4006a09e667f3bcd
 * QA-9: blaschkeDerivMagSq⁻¹(z=i, a=sqrt(λ₀)) = 1.016128772116579
 *   hex: 3ff042103e4c3ed8
 */

// ── Types ──────────────────────────────────────────────────────────────────────

/** Complex number as [real, imag] tuple. */
export type C = [number, number];

/** The IQuantumArith port interface. */
export interface IQuantumArith {
  readonly zero: C;
  readonly one: C;
  ofReal(r: number): C;
  ofPolar(r: number, theta: number): C;
  add(a: C, b: C): C;
  mul(a: C, b: C): C;
  conj(a: C): C;
  magSq(a: C): number;
  scale(s: number, a: C): C;
  neg(a: C): C;
  blaschke(z: C, a: C): C;
  blaschkeDerivMagSq(z: C, a: C): number;
  hlBumpParam(lambda0: number, theta: number): C;
  readonly invSqrt2: number;
  readonly tsirelsonS: number;
}

// ── Canonical implementation ───────────────────────────────────────────────────

const _zero: C = [0.0, 0.0];
const _one: C = [1.0, 0.0];
const _invSqrt2 = 1.0 / Math.sqrt(2.0);
const _tsirelsonS = 2.0 * Math.sqrt(2.0);

function _add(a: C, b: C): C { return [a[0] + b[0], a[1] + b[1]]; }
function _mul(a: C, b: C): C {
  return [a[0] * b[0] - a[1] * b[1], a[0] * b[1] + a[1] * b[0]];
}
function _conj(a: C): C { return [a[0], -a[1]]; }
function _magSq(a: C): number { return a[0] * a[0] + a[1] * a[1]; }
function _scale(s: number, a: C): C { return [s * a[0], s * a[1]]; }
function _neg(a: C): C { return [-a[0], -a[1]]; }
function _div(a: C, b: C): C {
  const d = _magSq(b);
  return [(a[0] * b[0] + a[1] * b[1]) / d, (a[1] * b[0] - a[0] * b[1]) / d];
}

function _blaschke(z: C, a: C): C {
  const num = _add(z, _neg(a));         // z − a
  const conjA = _conj(a);               // ā
  const conjAz = _mul(conjA, z);        // ā·z
  const den = _add(_one, _neg(conjAz)); // 1 − ā·z
  return _div(num, den);
}

function _blaschkeDerivMagSq(z: C, a: C): number {
  const aMagSq = _magSq(a);
  const conjA = _conj(a);
  const conjAz = _mul(conjA, z);
  const den = _add(_one, _neg(conjAz));
  const denSq = _magSq(den);
  const num = (1.0 - aMagSq) * (1.0 - aMagSq);
  return num / (denSq * denSq);
}

function _hlBumpParam(lambda0: number, theta: number): C {
  const r = Math.sqrt(lambda0);
  return [r * Math.cos(theta), r * Math.sin(theta)];
}

/** The canonical IQuantumArith implementation. Byte-lock reference. */
export const canonicalArith: IQuantumArith = {
  zero: _zero,
  one: _one,
  ofReal: (r) => [r, 0.0],
  ofPolar: (r, theta) => [r * Math.cos(theta), r * Math.sin(theta)],
  add: _add,
  mul: _mul,
  conj: _conj,
  magSq: _magSq,
  scale: _scale,
  neg: _neg,
  blaschke: _blaschke,
  blaschkeDerivMagSq: _blaschkeDerivMagSq,
  hlBumpParam: _hlBumpParam,
  invSqrt2: _invSqrt2,
  tsirelsonS: _tsirelsonS,
};

// ── Quantum gate helpers ───────────────────────────────────────────────────────

/** Born probability of a single amplitude: |amplitude|² */
export function bornProb(amplitude: C): number {
  return _magSq(amplitude);
}

/** Hadamard amplitude: H|k⟩ → amplitude for output bit outBit */
export function hadamardAmplitude(k: number, outBit: number): C {
  const sign = k === 1 && outBit === 1 ? -1.0 : 1.0;
  return [sign * _invSqrt2, 0.0];
}

// ── External adapter stubs ─────────────────────────────────────────────────────

/**
 * Qrisp adapter stub.
 * STUB: delegates to canonical until Qrisp Python bridge is integrated.
 * Replace each method with a Qrisp QuantumFloat call when ready.
 */
export const qrispAdapter: IQuantumArith = canonicalArith;

/**
 * OpenQASM 3.0 adapter stub.
 * STUB: delegates to canonical until OpenQASM toolchain is integrated.
 * Replace each method with an OpenQASM 3.0 circuit simulation call when ready.
 */
export const openQASMAdapter: IQuantumArith = canonicalArith;

/**
 * Emit an OpenQASM 3.0 circuit string for the Blaschke factor.
 * STUB: emits a placeholder circuit.
 */
export function emitBlaschkeCircuit(lambda0: number, theta: number, gridSize: number): string {
  const a = _hlBumpParam(lambda0, theta);
  return [
    `// OpenQASM 3.0 — Blaschke factor circuit (STUB)`,
    `// lambda0 = ${lambda0}, theta = ${theta}, grid_size = ${gridSize}`,
    `// a = ${a[0]} + ${a[1]}i`,
    `OPENQASM 3.0;`,
    `include "stdgates.inc";`,
    `qubit[${gridSize}] q;`,
    `// TODO: Solovay-Kitaev decomposition of Blaschke(z, a=${a[0]}+${a[1]}i)`,
  ].join("\n");
}

// ── Golden vector constants (for conformance tests) ────────────────────────────

export const GOLDEN_VECTORS = {
  QA1_ADD_REAL: 4.0,
  QA1_ADD_IMAG: 6.0,
  QA2_MUL_REAL: -5.0,
  QA2_MUL_IMAG: 10.0,
  QA3_MAG_SQ: 25.0,
  QA4_BLASCHKE_REAL: 0.47458659267475197,
  QA4_BLASCHKE_IMAG: 0.26034831334370395,
  QA5_BLASCHKE_DERIV_MAG_SQ: 1.1474510082682012,
  QA6_INV_SQRT2: 0.7071067811865476,
  QA7_BORN_PROB_BELL: 0.5,
  QA8_TSIRELSON_S: 2.8284271247461903,
  QA9_BLASCHKE_DERIV_INV: 1.0161287721165787,
} as const;

File: src/Core/QuantumArith.fs
namespace Zeta.Core

/// **QuantumArith — Canonical quantum arithmetic library (hexagonal port).**
///
/// This module defines the `IQuantumArith<'C>` port interface and provides
/// the canonical implementation using `Doubled<float>` (= Complex).
///
/// ## Hexagonal port pattern
///
/// The port interface `IQuantumArith<'C>` is the only thing the application
/// layer (WSet, HlAmplitudeEmu, ZSetISA) should depend on. External adapters
/// (OpenQASM, Qrisp) implement the same interface. The canonical implementation
/// is the byte-lock reference — all substrates (F#, TypeScript, Q#) must
/// produce the golden vectors in `testdata/quantum-arith-golden.json`.
///
/// ## Byte-lock discipline
///
/// All operations use IEEE 754 double precision (64-bit) with
/// round-to-nearest-even. No FMA, no fast-math, no platform-specific
/// intrinsics. The golden vectors are computed by the Python reference
/// in `docs/research/2026-08-09-quantum-arith-hexagonal-port-spec-lumen.md`.
///
/// ## Second quantum language
///
/// OpenQASM 3.0 is the recommended second quantum language (see spec doc).
/// The `OpenQASMAdapter` stub at the bottom of this file is the hexagonal
/// adapter that wraps OpenQASM 3.0 circuit output behind `IQuantumArith`.

open System

// ── Port interface ─────────────────────────────────────────────────────────────

/// The canonical quantum arithmetic port interface.
/// Parameterised over a complex number type 'C.
/// The canonical implementation uses Doubled<float> (= Complex).
type IQuantumArith<'C> =
    // Core complex arithmetic
    abstract Zero : 'C
    abstract One : 'C
    abstract OfReal : float -> 'C
    abstract OfPolar : r: float -> theta: float -> 'C
    abstract Add : 'C -> 'C -> 'C
    abstract Mul : 'C -> 'C -> 'C
    abstract Conj : 'C -> 'C
    abstract MagSq : 'C -> float
    abstract Scale : float -> 'C -> 'C
    abstract Neg : 'C -> 'C
    // Blaschke factor (canonical conformal map primitive)
    abstract Blaschke : z: 'C -> a: 'C -> 'C
    abstract BlascheDerivMagSq : z: 'C -> a: 'C -> float
    abstract HlBumpParam : lambda0: float -> theta: float -> 'C
    // Quantum gate layer (Born-probability)
    abstract InvSqrt2 : float
    abstract TsirelsonS : float

// ── Canonical implementation (Doubled<float> = Complex) ───────────────────────

/// Canonical QuantumArith implementation using Doubled<float> (= Complex).
/// This is the byte-lock reference. All golden vectors must match exactly.
[<RequireQualifiedAccess>]
module QuantumArith =

    // ── Internal helpers ────────────────────────────────────────────────────────

    /// Complex number as a pair of floats (real, imag).
    /// We use a plain tuple internally to avoid any boxing overhead.
    type C = float * float

    let re (z: C) = fst z
    let im (z: C) = snd z
    let mk r i : C = (r, i)

    // ── Core operations ─────────────────────────────────────────────────────────

    let zero : C = mk 0.0 0.0
    let one  : C = mk 1.0 0.0

    let ofReal (r: float) : C = mk r 0.0

    /// r·cos(θ) + r·sin(θ)·i
    let ofPolar (r: float) (theta: float) : C =
        mk (r * Math.Cos theta) (r * Math.Sin theta)

    /// (ar+br) + (ai+bi)i
    let add (a: C) (b: C) : C =
        mk (re a + re b) (im a + im b)

    /// (ar·br−ai·bi) + (ar·bi+ai·br)i
    let mul (a: C) (b: C) : C =
        mk (re a * re b - im a * im b) (re a * im b + im a * re b)

    /// ar − ai·i
    let conj (a: C) : C = mk (re a) (-(im a))

    /// ar² + ai²
    let magSq (a: C) : float = re a * re a + im a * im a

    /// (s·ar) + (s·ai)i
    let scale (s: float) (a: C) : C = mk (s * re a) (s * im a)

    /// (−ar) + (−ai)i
    let neg (a: C) : C = mk (-(re a)) (-(im a))

    /// Complex division: a / b
    let div (a: C) (b: C) : C =
        let d = magSq b
        mk ((re a * re b + im a * im b) / d) ((im a * re b - re a * im b) / d)

    // ── Blaschke factor ─────────────────────────────────────────────────────────

    /// Blaschke factor: (z − a) / (1 − ā·z)
    /// Requires |a| < 1 for the map to be well-defined on the unit disk.
    let blaschke (z: C) (a: C) : C =
        let num = add z (neg a)                   // z − a
        let conjA = conj a                         // ā
        let conjAz = mul conjA z                   // ā·z
        let den = add one (neg conjAz)             // 1 − ā·z
        div num den

    /// |(d/dz) blaschke(z, a)|² = (1 − |a|²)² / |1 − ā·z|⁴
    let blaschkeDerivMagSq (z: C) (a: C) : float =
        let aMagSq = magSq a
        let conjA = conj a
        let conjAz = mul conjA z
        let den = add one (neg conjAz)             // 1 − ā·z
        let denSq = magSq den
        let num = (1.0 - aMagSq) * (1.0 - aMagSq)
        num / (denSq * denSq)

    /// HL bump parameter: a = √λ₀ · e^{iθ}
    /// Places a INSIDE the unit disk (|a| = √λ₀ < 1 for λ₀ < 1).
    let hlBumpParam (lambda0: float) (theta: float) : C =
        ofPolar (Math.Sqrt lambda0) theta

    // ── Quantum constants ───────────────────────────────────────────────────────

    /// 1/√2 — the Hadamard amplitude.
    /// Golden vector QA-6: hex 3fe6a09e667f3bcc
    let invSqrt2 : float = 1.0 / Math.Sqrt 2.0

    /// 2√2 — the Tsirelson bound.
    /// Golden vector QA-8: hex 4006a09e667f3bcd
    let tsirelsonS : float = 2.0 * Math.Sqrt 2.0

    // ── Hadamard gate on a 2-state WSet ─────────────────────────────────────────

    /// H|0⟩ = (1/√2)|0⟩ + (1/√2)|1⟩
    /// H|1⟩ = (1/√2)|0⟩ − (1/√2)|1⟩
    let hadamardAmplitude (k: int) (outBit: int) : C =
        let sign = if k = 1 && outBit = 1 then -1.0 else 1.0
        ofReal (sign * invSqrt2)

    // ── Born probability ─────────────────────────────────────────────────────────

    /// Born probability of a single amplitude: |amplitude|²
    let bornProb (amplitude: C) : float = magSq amplitude

    // ── IQuantumArith instance ───────────────────────────────────────────────────

    /// The canonical IQuantumArith instance.
    let canonical : IQuantumArith<C> =
        { new IQuantumArith<C> with
            member _.Zero = zero
            member _.One = one
            member _.OfReal r = ofReal r
            member _.OfPolar r theta = ofPolar r theta
            member _.Add a b = add a b
            member _.Mul a b = mul a b
            member _.Conj a = conj a
            member _.MagSq a = magSq a
            member _.Scale s a = scale s a
            member _.Neg a = neg a
            member _.Blaschke z a = blaschke z a
            member _.BlascheDerivMagSq z a = blaschkeDerivMagSq z a
            member _.HlBumpParam lambda0 theta = hlBumpParam lambda0 theta
            member _.InvSqrt2 = invSqrt2
            member _.TsirelsonS = tsirelsonS }

    // ── Conversion to/from Doubled<float> (IStarRing<Complex>) ─────────────────

    /// Convert QuantumArith.C to Doubled<float> (the WSet weight type).
    let toDoubled (c: C) : Doubled<float> = Doubled.make (re c) (im c)

    /// Convert Doubled<float> to QuantumArith.C.
    let ofDoubled (d: Doubled<float>) : C = mk d.Real d.Imag

// ── OpenQASM 3.0 adapter stub ─────────────────────────────────────────────────

/// Hexagonal adapter stub for OpenQASM 3.0 circuit output.
/// This adapter wraps OpenQASM 3.0 circuit simulation behind IQuantumArith.
/// The canonical implementation is the byte-lock reference; this adapter
/// is for hardware execution and cross-validation only.
///
/// ## Replacement protocol
///
/// When OpenQASM 3.0 toolchain support matures:
/// 1. Implement each method by calling the OpenQASM 3.0 simulator
/// 2. Add the golden vector for each new operation
/// 3. Run the byte-lock CI gate — it must pass within 1 ULP
/// 4. The canonical implementation remains the byte-lock reference
///
/// ## Current status: STUB — all methods delegate to canonical
[<RequireQualifiedAccess>]
module OpenQASMAdapter =

    /// The OpenQASM 3.0 adapter. Currently delegates to canonical.
    /// Replace each method body with an OpenQASM 3.0 circuit call when ready.
    let adapter : IQuantumArith<QuantumArith.C> =
        // STUB: delegates to canonical until OpenQASM toolchain is integrated
        QuantumArith.canonical

    /// Emit an OpenQASM 3.0 circuit string for the Blaschke factor.
    /// This is the first step toward hardware execution.
    let emitBlaschkeCircuit (lambda0: float) (theta: float) (gridSize: int) : string =
        let a = QuantumArith.hlBumpParam lambda0 theta
        let ar = QuantumArith.toDoubled a |> fun d -> d.Real
        let ai = QuantumArith.toDoubled a |> fun d -> d.Imag
        // STUB: emit a placeholder OpenQASM 3.0 circuit
        // The real implementation would compile the Blaschke factor into
        // a sequence of Ry, Rz, CNOT gates using the Solovay-Kitaev algorithm.
        System.String.Format(
            "// OpenQASM 3.0 — Blaschke factor circuit (STUB)\n" +
            "// lambda0 = {0}, theta = {1}, grid_size = {2}\n" +
            "// a = {3} + {4}i\n" +
            "OPENQASM 3.0;\n" +
            "include \"stdgates.inc\";\n" +
            "qubit[{2}] q;\n" +
            "// TODO: Solovay-Kitaev decomposition of Blaschke(z, a={3}+{4}i)\n",
            lambda0, theta, gridSize, ar, ai)
